File: ios/OakApp/Networking/SSEClient.swift
import Foundation

/// Opens the chat stream (`POST /api/chat`) and turns the SSE byte stream into an
/// `AsyncThrowingStream<SSEEvent, Error>` (component-design.md "Networking layer").
///
/// It borrows ``OakAPIClient`` to attach the Bearer header + base URL (and to own
/// the `URLSession`), splits the response bytes into lines via ``ByteLineSplitter``
/// (NOT `URLSession.AsyncBytes.lines`, which drops the empty lines SSE needs to
/// delimit frames — see ``ByteLineSplitter``), feeds each line through ``SSEParser``,
/// and yields decoded events. Error contract (component-design.md "Interface
/// Definitions"):
///   * **Pre-stream HTTP failure** (rate limit, 413, 503, …): the non-2xx body is
///     read off the stream, mapped via ``OakError/validate(_:data:)``, and the
///     stream finishes by throwing that `OakError` before any event is yielded.
///   * **Transport drop mid-stream**: surfaces as a thrown `OakError.transport`.
///   * **SSE `error` event**: yielded as `.error(...)`, then the stream finishes
///     normally (transport faults ride this in-band event, per the wire contract).
///
/// The work runs in a child `Task` cancelled on stream termination, so a view that
/// disappears (or a new turn) tears down the connection.
struct SSEClient: Sendable {
  private let apiClient: OakAPIClient

  init(apiClient: OakAPIClient) {
    self.apiClient = apiClient
  }

  /// Opens the stream for one chat turn. The returned stream yields events until
  /// the terminal `answer`/`error` (or completion), then finishes.
  ///
  /// Chat works for guests too: `requiresAuth` here means "attach the Bearer token
  /// when signed in" (raises the rate limit + identity); a guest simply sends no
  /// Authorization header.
  func stream(_ request: ChatRequest) -> AsyncThrowingStream<SSEEvent, Error> {
    let endpoint = Endpoint(method: .post, path: "/api/chat", body: request, requiresAuth: true)
    return openEventStream(endpoint) { SSEParser() }
  }

  /// Opens the stream for one team-builder assistant turn (`POST /api/teams/assistant`).
  /// A sibling of ``stream(_:)`` over the SAME byte/frame plumbing (``openEventStream``)
  /// but a different parser: the builder's terminal `answer` carries a `BuilderAnswer`
  /// (not an `OakAnswer`) and there is no `scope` event. Signed-in only — a guest's
  /// missing Bearer surfaces as a `401` → `OakError.unauthorized` thrown before any
  /// event is yielded.
  func streamBuilder(
    _ request: TeamsAssistantRequest
  ) -> AsyncThrowingStream<BuilderSSEEvent, Error> {
    let endpoint = Endpoint(
      method: .post, path: "/api/teams/assistant", body: request, requiresAuth: true)
    return openEventStream(endpoint) { BuilderSSEParser() }
  }

  /// The shared SSE stream loop, generic over the frame parser (component-design.md
  /// "Networking layer"). It opens the byte stream, splits bytes into lines OURSELVES
  /// preserving empty lines (SSE delimits each event with a blank line
  /// `event: …\ndata: …\n\n`, which `bytes.lines`/`AsyncLineSequence` silently drops —
  /// collapsing every event's `data:` together and breaking the terminal frame's
  /// decode), feeds each line through `parser`, and yields decoded events. The error
  /// contract is uniform across families: a pre-stream HTTP failure throws an
  /// `OakError` out of `openByteStream` before any event is yielded; a mid-stream drop
  /// surfaces as `OakError.transportFailure`; a cancellation finishes cleanly.
  ///
  /// The work runs in a child `Task` cancelled on stream termination, so a view that
  /// disappears (or a new turn) tears down the connection.
  private func openEventStream<Parser: SSELineParser>(
    _ endpoint: Endpoint,
    makeParser: @escaping @Sendable () -> Parser
  ) -> AsyncThrowingStream<Parser.Event, Error> {
    let apiClient = self.apiClient
    return AsyncThrowingStream { continuation in
      let task = Task {
        do {
          let bytes = try await apiClient.openByteStream(endpoint)

          var parser = makeParser()
          var splitter = ByteLineSplitter()
          for try await byte in bytes {
            guard let line = splitter.consume(byte) else { continue }
            for event in try parser.consume(line: line) {
              continuation.yield(event)
            }
          }
          // Flush a final, unterminated line (a stream that doesn't end in `\n`).
          if let line = splitter.finish() {
            for event in try parser.consume(line: line) {
              continuation.yield(event)
            }
          }
          for event in try parser.finish() {
            continuation.yield(event)
          }
          continuation.finish()
        } catch let urlError as URLError where urlError.code == .cancelled {
          continuation.finish()
        } catch is CancellationError {
          continuation.finish()
        } catch let error as OakError {
          continuation.finish(throwing: error)
        } catch {
          continuation.finish(throwing: OakError.transportFailure(error))
        }
      }
      continuation.onTermination = { _ in
        task.cancel()
      }
    }
  }
}

/// Splits a raw byte stream into lines, **preserving empty lines** — the piece SSE
/// needs that `URLSession.AsyncBytes.lines` gets wrong.
///
/// `AsyncLineSequence` (what `bytes.lines` yields) silently drops empty lines. SSE
/// delimits each event with a blank line (`event: …\ndata: …\n\n`), and
/// ``SSEParser`` relies on that blank line to dispatch a frame; without it the
/// parser concatenates every event's `data:` payload and the terminal `answer`
/// fails to decode (the whole chat stream breaks). This splitter emits a line on
/// every `\n` (`0x0A`) — including the empty string between consecutive newlines.
///
/// Splitting on the ASCII byte `0x0A` is UTF-8-safe: `0x0A` never appears inside a
/// multi-byte UTF-8 sequence, so each accumulated line decodes whole. A trailing
/// `\r` (CRLF input) is intentionally left on the line for ``SSEParser`` to strip.
///
/// Pure and synchronous (no I/O, no actor), so it's trivially unit-testable against
/// recorded byte buffers.
struct ByteLineSplitter {
  private var buffer = [UInt8]()

  /// Feeds one byte. Returns the completed line when `byte` is a newline (`0x0A`),
  /// including an empty string for a blank line; returns `nil` otherwise.
  mutating func consume(_ byte: UInt8) -> String? {
    if byte == 0x0A {
      let line = String(decoding: buffer, as: UTF8.self)
      buffer.removeAll(keepingCapacity: true)
      return line
    }
    buffer.append(byte)
    return nil
  }

  /// Returns any buffered final line not terminated by a newline, or `nil` when the
  /// stream ended cleanly on a `\n`.
  mutating func finish() -> String? {
    guard !buffer.isEmpty else { return nil }
    let line = String(decoding: buffer, as: UTF8.self)
    buffer.removeAll(keepingCapacity: true)
    return line
  }
}

File: android/app/src/main/kotlin/ai/gowtam/oak/networking/SseClient.kt
package ai.gowtam.oak.networking

import ai.gowtam.oak.wire.BuilderSseEvent
import ai.gowtam.oak.wire.ChatRequest
import ai.gowtam.oak.wire.SseEvent
import ai.gowtam.oak.wire.TeamsAssistantRequest
import java.io.IOException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import okhttp3.Response
import okio.Buffer

/**
 * Opens the chat/team-builder byte streams and turns them into cold `Flow`s of decoded
 * events (component-design.md "Networking layer"). Borrows [OakApiClient] for the Bearer
 * header + base URL + the pre-stream error mapping in [OakApiClient.openByteStream],
 * splits the response bytes into lines via [ByteLineSplitter] (NOT a line reader that
 * drops the empty lines SSE needs to delimit frames — see [ByteLineSplitter]'s doc),
 * feeds each line through the endpoint's [SseLineParser], and emits decoded events.
 *
 * Error contract (api-usage.md "Error mapping"):
 *  - **Pre-stream HTTP failure** (rate limit, 413, 503, …): already mapped and thrown by
 *    [OakApiClient.openByteStream] before this class sees a response — so it surfaces
 *    before any event is emitted.
 *  - **Transport drop mid-stream**: a body-read `IOException` is caught and rethrown as
 *    [OakError.Transport].
 *  - **SSE `error` event**: yielded as [SseEvent.Error]/[BuilderSseEvent.Error], then the
 *    flow completes normally (transport faults ride this in-band event only when the
 *    server itself detected them; a raw connection drop is the case above).
 *
 * The returned `Flow` is cold (nothing runs until collected) and cancellable: cancelling
 * the collecting coroutine closes the underlying OkHttp response/call promptly, even
 * while a blocking body read is in flight, via a completion handler registered on the
 * upstream job (see [openEventStream]).
 */
class SseClient(private val apiClient: OakApiClient) {

    /** Opens the stream for one chat turn (`POST /api/chat`, guest or signed-in). */
    fun stream(request: ChatRequest): Flow<SseEvent> {
        val endpoint = Endpoint(
            method = Endpoint.Method.POST,
            path = "/api/chat",
            body = Endpoint.jsonBody(ChatRequest.serializer(), request),
            requiresAuth = true,
        )
        return openEventStream(endpoint) { SseParser() }
    }

    /**
     * Reattaches to a durable turn's live stream
     * (`GET /api/chat/turns/:id/stream?session_id=…`; background-turns/design.md §4).
     * The server replays the turn's full buffered event list from the start (the
     * client rebuilds its in-flight UI on the `turn` frame) then tails live until a
     * terminal event; a turn that is already terminal ends the replay with its
     * terminal event and closes — so reattach-mid-flight and reattach-after-completion
     * are one code path. An unknown/expired turn is a pre-stream `404` →
     * [OakError.Http] thrown before any event (the caller treats it as a dead turn).
     */
    fun resume(turnId: String, sessionId: String): Flow<SseEvent> {
        val endpoint = Endpoint(
            method = Endpoint.Method.GET,
            path = "/api/chat/turns/$turnId/stream",
            queryItems = listOf("session_id" to sessionId),
            requiresAuth = true,
        )
        return openEventStream(endpoint) { SseParser() }
    }

    /**
     * Explicitly stops a durable turn (`POST /api/chat/turns/:id/stop`;
     * background-turns/design.md §4 / BT-4): the server aborts generation, the turn
     * transitions to `stopped`, and nothing is persisted. A non-streaming JSON call
     * (borrows [OakApiClient] like the streams do): the Bearer header identifies a
     * signed-in owner; a guest is authorized by the `{ session_id }` body. Stopping
     * an unknown or already-terminal turn is a server-side no-op. Failures surface as
     * an [OakError] — the caller tears down its local stream regardless.
     */
    suspend fun stop(turnId: String, sessionId: String) {
        val endpoint = Endpoint(
            method = Endpoint.Method.POST,
            path = "/api/chat/turns/$turnId/stop",
            body = Endpoint.jsonBody(StopTurnBody.serializer(), StopTurnBody(sessionId)),
            requiresAuth = true,
        )
        apiClient.sendNoContent(endpoint)
    }

    /**
     * Opens the stream for one team-builder assistant turn (`POST /api/teams/assistant`,
     * signed-in only — a guest's missing Bearer surfaces as a `401` →
     * [OakError.Unauthorized] thrown before any event).
     */
    fun streamBuilder(request: TeamsAssistantRequest): Flow<BuilderSseEvent> {
        val endpoint = Endpoint(
            method = Endpoint.Method.POST,
            path = "/api/teams/assistant",
            body = Endpoint.jsonBody(TeamsAssistantRequest.serializer(), request),
            requiresAuth = true,
        )
        return openEventStream(endpoint) { BuilderSseParser() }
    }

    /**
     * The shared SSE stream loop, generic over the frame parser. Opens the byte stream,
     * splits bytes into lines ourselves preserving empty lines, feeds each line through
     * `parser`, and emits decoded events; flushes a trailing unterminated line
     * ([ByteLineSplitter.finish]) and the parser's own end-of-stream flush
     * ([SseLineParser.finish]) once the body is exhausted.
     *
     * Runs on [Dispatchers.IO] since the body read is a blocking Okio call. A completion
     * handler on the upstream job closes the OkHttp [Response] the instant the flow is
     * cancelled — a plain `try/finally` alone would only close it once the blocking read
     * unblocks naturally (i.e. once the next chunk arrives), which could be never.
     */
    private fun <T> openEventStream(
        endpoint: Endpoint,
        makeParser: () -> SseLineParser<T>,
    ): Flow<T> = flow {
        val response = apiClient.openByteStream(endpoint)
        val closeOnCancel = currentCoroutineContext()[Job]?.invokeOnCompletion { response.close() }
        try {
            emitFrom(response, makeParser())
        } catch (e: IOException) {
            throw OakError.transportFailure(e)
        } finally {
            closeOnCancel?.dispose()
            response.close()
        }
    }.flowOn(Dispatchers.IO)

    private suspend fun <T> kotlinx.coroutines.flow.FlowCollector<T>.emitFrom(
        response: Response,
        parser: SseLineParser<T>,
    ) {
        val source = response.body?.source() ?: throw OakError.Transport("empty_body")
        val splitter = ByteLineSplitter()
        val readBuffer = Buffer()
        while (true) {
            val read = source.read(readBuffer, CHUNK_BYTES)
            if (read == -1L) break
            val chunk = readBuffer.readByteArray()
            for (line in splitter.consume(chunk)) {
                for (event in parser.consume(line)) emit(event)
            }
        }
        splitter.finish()?.let { line ->
            for (event in parser.consume(line)) emit(event)
        }
        for (event in parser.finish()) emit(event)
    }

    private companion object {
        const val CHUNK_BYTES = 8192L
    }
}

/** `POST /api/chat/turns/:id/stop` body — the guest ownership proof (ignored for a
 * signed-in caller, who is identified by the Bearer token). */
@Serializable
private data class StopTurnBody(@SerialName("session_id") val sessionId: String)
